README.md
Socket Address
==============

An [Elixir][] convenience library for manipulating Internet socket addresses.

[![Build Status](https://travis-ci.org/elasticdog/socket_address.svg?branch=master)](https://travis-ci.org/elasticdog/socket_address)
[![Coverage Status](https://coveralls.io/repos/github/elasticdog/socket_address/badge.svg?branch=master)](https://coveralls.io/github/elasticdog/socket_address?branch=master)
[![Hex.pm](https://img.shields.io/hexpm/v/socket_address.svg)](https://hex.pm/packages/socket_address)

[Elixir]: http://elixir-lang.org/

Usage
-----

```elixir
iex> {:ok, socket_address} = SocketAddress.new("127.0.0.1", 80)
iex> socket_address
#SocketAddress<127.0.0.1:80>
iex> socket_address.ip
{127, 0, 0, 1}
iex> socket_address.port
80
iex> SocketAddress.to_opts(socket_address, [compress: true])
[ip: {127, 0, 0, 1}, port: 80, compress: true]

iex> {:ok, socket_address} = SocketAddress.new("fe80::204:acff:fe17:bf38", 80)
iex> socket_address
#SocketAddress<[FE80::204:ACFF:FE17:BF38]:80>
iex> socket_address.ip
{65152, 0, 0, 0, 516, 44287, 65047, 48952}

iex> SocketAddress.new("100.200.300.400", 80)
{:error, :invalid_ip}

iex> SocketAddress.new("0.0.0.0", 99999)
{:error, :invalid_port}
```

### Installation

Releases are published to [Hex](https://hex.pm) and can be installed by adding
`socket_address` to your list of dependencies in `mix.exs`:

```elixir
defp deps do
  [{:socket_address, "~> 0.1"}]
end
```

### Documentation

Generated [ExDoc][] API reference documentation can be found at
<https://hexdocs.pm/socket_address/>

[ExDoc]: https://github.com/elixir-lang/ex_doc

Background
----------

A *socket address* is the combination of an IP address and a port number. This
can be used along with a transport protocol in order to define an endpoint of
a connection across a network.

### The Problem

When using the Erlang HTTP server [Cowboy][], either directly or indirectly via
[Plug][] / [Phoenix][] / etc., if you want to bind the listener to a specific
IP address, you must pass the address in a **tuple format**. For example, if
you want to bind to the localhost address of _127.0.0.1_, you must pass in the
tuple:

    {127, 0, 0, 1}

The required tuple format gets more confusing if you want to bind to an IPv6
address. For instance, the IPv6 address _FE80::204:ACFF:FE17:BF38_ must be
passed in as the tuple:

    {65152, 0, 0, 0, 516, 44287, 65047, 48952}

[Cowboy]: https://github.com/ninenines/cowboy
[Plug]: https://github.com/elixir-lang/plug
[Phoenix]: http://www.phoenixframework.org/

### The Solution

The Socket Address library eases the mental burden of using Cowboy's tuple
format by generating it for you based on normal IP address strings. It sanity
checks both the IP address and port number you give, and then packages it into
a struct for use when setting your listener options.

It's essentially a thin wrapper around the Erlang [`:inet.parse_address/1`][]
function, with some added niceties.

[`:inet.parse_address/1`]: http://erlang.org/doc/man/inet.html#parse_address-1

Contributing
------------

This project welcomes contributions from everyone. If you're thinking of
helping out, please read the [guidelines for contributing][contributing].

[contributing]: https://github.com/elasticdog/socket_address/blob/master/CONTRIBUTING.md

License
-------

Socket Address is provided under the terms of the
[ISC License](https://en.wikipedia.org/wiki/ISC_license).

Copyright &copy; 2016-2017, [Aaron Bull Schaefer](mailto:aaron@elasticdog.com).


CHANGELOG.md
Changelog for Socket Address
============================

All notable changes to the project will be documented in this file.

[HEAD][] / unreleased
---------------------

[v0.2.0][] / 2017-01-03
-----------------------

### Added

* `SocketAddress.to_opts/2` helper function for generating Cowboy options.
* Repository metadata so documentation will automatically link to source files.

[v0.1.0][] / 2016-12-30
-----------------------

* Initial project release.

[HEAD]: https://github.com/elasticdog/socket_address/compare/v0.2.0...master
[v0.2.0]: https://github.com/elasticdog/socket_address/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/elasticdog/socket_address/tree/v0.1.0


test/socket_address_test.exs
defmodule SocketAddressTest do
  use ExUnit.Case, async: true
  doctest SocketAddress
end


mix.exs
defmodule SocketAddress.Mixfile do
  use Mix.Project

  def project() do
    [
      app: :socket_address,
      version: "0.2.0",
      elixir: "~> 1.3",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),

      # Docs
      name: "Socket Address",
      source_url: "https://github.com/elasticdog/socket_address",
      homepage_url: "https://github.com/elasticdog/socket_address",
      docs: [
        main: "readme",
        extras: [
          "README.md": [title: "README"],
          "CONTRIBUTING.md": [title: "Contributing"],
        ],
      ],

      # Hex.pm
      description: description(),
      package: package(),

      # Tests
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [coveralls: :test],
    ]
  end

  def application() do
    [applications: [:logger]]
  end

  defp deps() do
    [
      {:credo, "~> 0.5", only: :dev},
      {:dialyxir, "~> 0.4", only: :dev},
      {:ex_doc, "~> 0.14", only: :dev},

      {:excoveralls, "~> 0.5", only: :test},
    ]
  end

  defp description() do
    """
    An Elixir convenience library for manipulating Internet socket addresses.
    """
  end

  defp package() do
    [
      licenses: ["ISC"],
      maintainers: ["Aaron Bull Schaefer"],
      links: %{"GitHub" => "https://github.com/elasticdog/socket_address"}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :socket_address, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:socket_address, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/socket_address.ex
defmodule SocketAddress do
  @moduledoc """
  Defines an Internet socket address.

  The SocketAddress struct contains the fields `ip` (stored as a tuple) and
  `port`, which can be accessed directly:

      iex> {:ok, socket_address} = SocketAddress.new("127.0.0.1", 80)
      iex> socket_address
      #SocketAddress<127.0.0.1:80>
      iex> socket_address.ip
      {127, 0, 0, 1}
      iex> socket_address.port
      80
  """

  @typedoc "The IP address of a socket"
  @type ip_address :: String.t | char_list | :inet.ip_address

  @typedoc "The port number of a socket"
  @type port_number :: 0..65_535

  @typedoc "The socket address type"
  @type t :: %SocketAddress{ip: :inet.ip_address, port: port_number}

  @enforce_keys [:ip, :port]
  defstruct [:ip, :port]

  @valid_ports 0..65_535

  @doc """
  Creates a new socket address with the given `ip` and `port`.

  Returns `{:ok, socket_address}` if the IP address and port number are valid,
  returns `{:error, reason}` otherwise. A valid IP address is an IPv4 or IPv6
  address that can be parsed by `:inet.parse_address/1`, and a valid port must
  be an integer in the range of `#{inspect @valid_ports}`.

  ## Examples

      iex> {:ok, socket_address} = SocketAddress.new("127.0.0.1", 80)
      iex> socket_address
      #SocketAddress<127.0.0.1:80>
      iex> {:ok, socket_address} = SocketAddress.new('127.0.0.1', 80)
      iex> socket_address
      #SocketAddress<127.0.0.1:80>
      iex> {:ok, socket_address} = SocketAddress.new({127, 0, 0, 1}, 80)
      iex> socket_address
      #SocketAddress<127.0.0.1:80>

      iex> {:ok, socket_address} = SocketAddress.new("fe80::204:acff:fe17:bf38", 80)
      iex> socket_address
      #SocketAddress<[FE80::204:ACFF:FE17:BF38]:80>

      iex> SocketAddress.new("100.200.300.400", 80)
      {:error, :invalid_ip}

      iex> SocketAddress.new("0.0.0.0", 99999)
      {:error, :invalid_port}
  """
  @spec new(ip_address, port_number) ::
    {:ok, t} |
    {:error, :invalid_ip | :invalid_port}
  def new(ip, port) do
    ip_address = parse(ip)

    cond do
      ip_address == nil ->
        {:error, :invalid_ip}
      Enum.member?(@valid_ports, port) == false ->
        {:error, :invalid_port}
      true ->
        {:ok, %SocketAddress{ip: ip_address, port: port}}
    end
  end

  # Parses an `t:ip_address/0` into an `t::inet.ip_address/0` tuple.
  #
  # Returns `nil` if there was an error with parsing.
  @spec parse(ip_address) :: :inet.ip_address | nil
  defp parse(ip_address) when is_binary(ip_address) do
    ip_address |> String.to_char_list |> parse
  end
  defp parse(ip_address) when is_list(ip_address) do
    case :inet.parse_address(ip_address) do
      {:ok, value} -> value
      {:error, :einval} -> nil
    end
  end
  defp parse(ip_address) when is_tuple(ip_address) do
    try do
      :inet.ntoa(ip_address)
    rescue
      ArgumentError -> nil
    else
      {:error, :einval} -> nil
      _ -> ip_address
    end
  end
  defp parse(_), do: nil

  @doc """
  Converts a socket address to an options keyword list.

  Returns a keyword list of the socket address `ip` and `port` fields merged
  with any provided `opts`. All keys, including duplicated keys, given in
  `opts` will be added to the socket address fields, overriding any existing
  one.

  ## Examples

      iex> {:ok, socket_address} = SocketAddress.new("127.0.0.1", 80)
      iex> SocketAddress.to_opts(socket_address)
      [ip: {127, 0, 0, 1}, port: 80]

      iex> {:ok, socket_address} = SocketAddress.new("127.0.0.1", 80)
      iex> SocketAddress.to_opts(socket_address, [compress: true])
      [ip: {127, 0, 0, 1}, port: 80, compress: true]

      iex> {:ok, socket_address} = SocketAddress.new("127.0.0.1", 80)
      iex> SocketAddress.to_opts(socket_address, [port: 8888])
      [ip: {127, 0, 0, 1}, port: 8888]
  """
  @spec to_opts(t, Keyword.t) :: Keyword.t
  def to_opts(%SocketAddress{ip: ip, port: port}, opts \\ []) do
    Keyword.merge([ip: ip, port: port], opts)
  end
end


defimpl Inspect, for: SocketAddress do
  import Inspect.Algebra

  def inspect(socket_address, _opts) do
    surround("#SocketAddress<", "#{socket_address}", ">")
  end
end


defimpl String.Chars, for: SocketAddress do
  def to_string(socket_address) do
    case tuple_size(socket_address.ip) do
      4 -> "#{:inet.ntoa(socket_address.ip)}:#{socket_address.port}"
      8 -> "[#{:inet.ntoa(socket_address.ip)}]:#{socket_address.port}"
    end
  end
end
